### PR TITLE
feat: add learn layout

### PR DESCRIPTION
**Pull request by ovflowd**
## Description

This PR introduces the Learn Layout for the Node.js Website Redesign

## Validation

Layout looks as expected on Figma

### CODE REVIEW

File: components/Common/ProgressionSideBar/ProgressionSideBarIcon/index.tsx [hunk 1]
```diff
@@ -13,4 +13,4 @@ const ProgressionSidebarIcon: FC<SVGAttributes<SVGSVGElement>> = props => (
   </svg>
 );
 
-export default ProgressionSidebarIcon;
+export default ProgressionSideBarIcon;
```
Comment (bmuenzenmeyer) [inline components/Common/ProgressionSideBar/ProgressionSideBarIcon/index.tsx]: this seems like an arbitrary change

- it's lowercase in the figma
- it's a compound single word according to https://www.merriam-webster.com/dictionary/sidebar
- The entirety of the `Sidebar` component family is lowercased
- the rename didn't carry through to the usage within components/Common/ProgressionSideBar/ProgressionSideBarItem/index.tsx

`import ProgressionSidebarIcon`

I'm like [5 outta 10](https://capwatkins.com/blog/the-sliding-scale-of-giving-a-fuck) on changing it, but am concerned its not a complete rename. Note the filename vs the variable names.

image

File: components/withProgressionSidebar.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  context,
+}) => {
+  const { getSideNavigation } = useSiteNavigation();
+
+  const [[, sidebarNavigation]] = getSideNavigation([navKey], context);
+
+  const mappedProgressionSideBarItems = sidebarNavigation.items.map(
+    ([, { label, items }]) => ({
+      groupName: label,
+      items: items.map(([, { label, link }]) => ({ title: label, url: link })),
```
Comment (bmuenzenmeyer) [inline components/withProgressionSidebar.tsx]: I could see people getting confused by the double scoping of the variable `label` here
```suggestion
      items: items.map(([, { itemLabel, link }]) => ({ title: itemLabel, url: link })),
```

File: components/Common/ActiveLink/index.tsx [hunk 3]
```diff
@@ -23,7 +23,10 @@ const ActiveLink: FC<ActiveLocalizedLinkProps> = ({
 
   const finalClassName = classNames(className, {
     [activeClassName]: allowSubPath
-      ? pathname.startsWith(href.toString())
+      ? // When using allowSubPath we want only to check if
+        // the current pathname starts with the utmost upper level
+        // of an href (e.g. /docs/...)
+        pathname.startsWith(`/${href.toString().split('/')[1]}`)
       : href.toString() === pathname,
   });
```
Comment (bmuenzenmeyer) [inline components/Common/ActiveLink/index.tsx:31]: wondering if this reaches the level of complexity to warrant a unit test

**Review by bmuenzenmeyer — COMMENTED** [review]
looking good so far - small nits only

**Review by canerakdas — COMMENTED** [review]
When I checked from [Figma](https://www.figma.com/file/pu1vZPqNIM7BePd6W8APA5/Node.js?node-id=422%3A7921&mode=dev);

- `Crosslink` component should be used after MDX content
- The `Select` component should be used instead of `ProgressionSidebar` on mobile resolution

**Comment by ovflowd** [comment]
> The Select component should be used instead of ProgressionSidebar on mobile resolution

I think that never got implemented. Let me do that.

> Crosslink component should be used after MDX content

Right

File: components/Common/ProgressionSideBar/ProgressionSideBarIcon/index.tsx (hunk 1 above)
Comment (ovflowd) [reply components/Common/ProgressionSideBar/ProgressionSideBarIcon/index.tsx]: Hmm, right it should be Sidebar, will do the change.

**Comment by ovflowd** [comment]
> When I checked from [Figma](https://www.figma.com/file/pu1vZPqNIM7BePd6W8APA5/Node.js?node-id=422%3A7921&mode=dev);
> 
> * `Crosslink` component should be used after MDX content
> * The `Select` component should be used instead of `ProgressionSidebar` on mobile resolution

I've added the missing parts.

**Comment by ovflowd** [comment]
image

File: components/Common/ProgressionSidebar/index.module.css [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    xs:border-r-0;
+
+  > section {
+    @apply flex
+      xs:hidden;
+  }
+
+  > div {
+    @apply hidden
+      w-full xs:flex;
```
Comment (canerakdas) [inline components/Common/ProgressionSidebar/index.module.css]: Only one Tailwind class per line 👀

**Review by canerakdas — APPROVED** [review]
LGTM!

I think instead of using the `mdxContent` class, we can turn it into a container component and use with CSS module

Similarly, for the `Crosslinks` in `withCrossLinks.tsx`, we can create a component called `CrosslinkRoot` in the `Crosslink` directory. With this approach, we can avoid inline styling

@ovflowd  If these changes sound good, I can create a new PR or update this PR

**Comment by ovflowd** [comment]
> I think instead of using the mdxContent class, we can turn it into a container component and use with CSS module

I did this in a commit, and then realized we cannot and should not do that. Well, are you saying to apply the styles directly on the `mdxRenderer`?

Well there's a major reason why I did that, if the "div" or container that encompasses `mdxContent` is directly on `mdxRenderer` we cannot do the CrossLinks on Learn layout with the current grid system. Mostly because the CrossLink components should be inside the div of `mdxContent`...

That's also why I moved the class of "mdxContent" to somewhere else, But making a dedicated component just for that... Could probably work... I'll see what I can do.

**Comment by ovflowd** [comment]
> Similarly, for the Crosslinks in withCrossLinks.tsx, we can create a component called CrosslinkRoot in the Crosslink directory. With this approach, we can avoid inline styling

Well, IDK, I know we avoid using class names directly within components, but that's to **style** a component. `withCrossLinks` is not a component, but a "provider", it "connects data" with components; It's pretty much what actually composes a component with something.

Hence I thought, well, it's fine in this case, rather than making yet another component CrossLinkRoot, or "CrossLinkGroups", but I'm fine revisiting this in the future if needed or if you or someone else have strong opinions about it.

**Comment by ovflowd** [comment]
> > I think instead of using the mdxContent class, we can turn it into a container component and use with CSS module
> 
> I did this in a commit, and then realized we cannot and should not do that. Well, are you saying to apply the styles directly on the `mdxRenderer`?
> 
> Well there's a major reason why I did that, if the "div" or container that encompasses `mdxContent` is directly on `mdxRenderer` we cannot do the CrossLinks on Learn layout with the current grid system. Mostly because the CrossLink components should be inside the div of `mdxContent`...
> 
> That's also why I moved the class of "mdxContent" to somewhere else, But making a dedicated component just for that... Could probably work... I'll see what I can do.

Also the `mdxContent` class name is used across many places (stories, components, etc)... I feel that this is the sort of container that makes more sense to be behind a global className, such as tailwind.

**Review by aymen94 — APPROVED** [review]

**Pull request merged**